### ISSUE TITLE
Running "berks update" with two sources fails with "undefined method `name'" error

### ISSUE THREAD

**Issue by StephenKing** [berkshelf/berkshelf#1676]
This is a follow-up to #1637.

I have the following `Berksfile` in [this cookbook](https://github.com/TYPO3-cookbooks/site-webhosttypo3org/tree/68c119856dfed320009f7fd0ac88ef1eec291abd):

```ruby
source "http://chef.typo3.org:26200"
source "https://supermarket.chef.io"

metadata

solver :ruby, :required
```

I get the following exception:

```
$ berks update
Resolving cookbook dependencies...
Fetching 'site-webhosttypo3org' from source at .
Fetching cookbook index from https://supermarket.chef.io...
Fetching cookbook index from http://chef.typo3.org:26200...
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:80:in `block in resolve': undefined method `name' for nil:NilClass (NoMethodError)
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:79:in `each'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:79:in `inject'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:79:in `resolve'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/solve-3.0.1/lib/solve.rb:64:in `it!'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/resolver.rb:78:in `resolve'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/installer.rb:175:in `install_from_universe'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/installer.rb:39:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/berksfile.rb:422:in `install'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/berksfile.rb:447:in `update'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/cli.rb:163:in `update'
	from /Users/st/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /Users/st/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/st/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/cli.rb:52:in `dispatch'
	from /Users/st/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/lib/berkshelf/cli.rb:27:in `execute!'
	from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/berkshelf-5.2.0/bin/berks:5:in `<top (required)>'
	from /usr/local/bin/berks:49:in `load'
	from /usr/local/bin/berks:49:in `<main>'
```
You can access the `/universe` endpoint at `http://chef.typo3.org:26200` publicly.

Versions:
```
Chef Development Kit Version: 1.0.3
berks version: 5.2.0
```

If you need any more information to reproduce it, feel free to ask. I'm willing to share anything.

**Comment by StephenKing** [comment]
@thommay this was about what I was talking today.

**Comment by northrup** [comment]
I'm getting the same error as well, have you made any progress on resolving this?

**Comment by StephenKing** [comment]
@northrup nope, didn't debug this any further. I'm always commenting out single endpoints (very annoying). It's not possible to install other berkshelf versions into ChefDK, which makes it pretty hard to bisect.

**Comment by northrup** [comment]
@StephenKing how would I debug this error to find which cookbook was throwing itself into a fit? 

For example I have the following going on right now:

```
D, [2017-02-21T01:45:40.265777 #7768] DEBUG -- : Using 'chefignore' at '/Users/jjn/.berkshelf/cookbooks/packagecloud-0.0.19/chefignore'
D, [2017-02-21T01:45:40.277353 #7768] DEBUG -- : Using 'chefignore' at '/Users/jjn/.berkshelf/cookbooks/packagecloud-0.2.5/chefignore'
D, [2017-02-21T01:45:40.468206 #7768] DEBUG -- : Using 'chefignore' at '/Users/jjn/.berkshelf/cookbooks/redisio-2.5.0/chefignore'
/Users/jjn/.rvm/gems/ruby-2.3.1/gems/solve-3.1.0/lib/solve/ruby_solver.rb:80:in `block in resolve': undefined method `name' for nil:NilClass (NoMethodError)
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/solve-3.1.0/lib/solve/ruby_solver.rb:79:in `each'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/solve-3.1.0/lib/solve/ruby_solver.rb:79:in `inject'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/solve-3.1.0/lib/solve/ruby_solver.rb:79:in `resolve'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/solve-3.1.0/lib/solve.rb:64:in `it!'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/resolver.rb:78:in `resolve'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/installer.rb:176:in `install_from_universe'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/installer.rb:40:in `run'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/berksfile.rb:422:in `install'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/berksfile.rb:447:in `update'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/cli.rb:163:in `update'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/cli.rb:52:in `dispatch'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/lib/berkshelf/cli.rb:27:in `execute!'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/gems/berkshelf-5.6.2/bin/berks:5:in `<top (required)>'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/bin/berks:23:in `load'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/bin/berks:23:in `<main>'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `eval'
        from /Users/jjn/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by northrup** [comment]
Or was your solution to just comment out the source, run the berks update, and then re-add the second source?

**Comment by StephenKing** [comment]
exactly..

**Comment by mrjcleaver** [comment]
:(

**Comment by chrisgit** [comment]
The error occurs in the solver, while it's not pretty there is a "work around" on the first post in the Github page of the solver https://github.com/berkshelf/solve/issues/64

NB: This isn't an official fix but it appears to work.

Downgrading Berkshelf to and older version (5.1.0) is possible but not that easy due to version constraints.

**Comment by northrup** [comment]
@chrisgit changing my `solver` as mentioned in the issue you cross-referenced fixed the problem for me.

**Comment by StephenKing** [comment]
I had trouble with the gecode solver before, that's why I switched to the ruby solver (`berks update` freezes often with multiple sources).
Good to hear that this workaround helps in your case, but it's not a real "fix" ;-)

**Comment by stucki** [comment]
Seems like this is still broken. Here's the workaround: https://github.com/berkshelf/solve/issues/64#issuecomment-277067851

**Comment by lamont-granquist** [comment]
appears that this was resolved via solve 3.1.1
